tests: workqueue tests overhaul adding new tests

maksimmasalski opened the pull request:
During inspection of the workqueue tests, I find out testing gaps.
Decided to add new test cases that can improve Zephyr OS testing
quality.
Added new test cases:
1. test_work_item_supplied_with_func
In docs described that work item supplied with a handler function,
prove that it works.
2. test_process_work_items_fifo
Test that system process work items in first-in, first-out manner.
3. test_sched_delayed_work_item
Verify that delayed work item processed after specific period of time
stated by user.
4. test_workqueue_max_number
Test the limit of number of workqueues created
5. test_cancel_processed_work_item Created test to increase branch
coverage.

Modified existing test cases:
1. test_work_submit_handler updated Doxygen tag, added more detailed
description"

Signed-off-by: Maksim Masalski <maksim.masalski@intel.com>

Commit:
tests: workqueue tests overhaul adding new tests

During inspection of the workqueue tests, I find out testing gaps.
Decided to add new test cases that can improve Zephyr OS testing
quality.
Added new test cases:
1. test_work_item_supplied_with_func
In docs described that work item supplied with a handler function,
prove that it works.
2. test_process_work_items_fifo
Test that system process work items in first-in, first-out manner.
3. test_sched_delayed_work_item
Verify that delayed work item processed after specific period of time
stated by user.
4. test_workqueue_max_number
Test the limit of number of workqueues created
5. test_cancel_processed_work_item Created test to increase branch
coverage.

Modified existing test cases:
1. test_work_submit_handler updated Doxygen tag, added more detailed
description"

Signed-off-by: Maksim Masalski <maksim.masalski@intel.com>

andyross commented on tests/kernel/workq/work_queue_api/src/main.c:
```diff
@@ -37,6 +50,154 @@ static struct k_sem sync_sema;
+ * - Test defines and initializes max number of the workqueues and starts them.
+ * @ingroup kernel_workqueue_tests
+ */
+void test_workqueue_max_number(void)
```
I still think this kind of requirement-chasing is silly, but this is as good a way to test this as any.

FWIW: this is likely to be fragile.  Setting MAX_WORK_Q_NUMBER high enough to be nontrivial means using a lot of memory for all those stacks.  We have a bunch of small devices in the CI test suite that are going to fail to link if you stuff too much junk into a single test.  This may need to be a kconfig and be set per-platform, or autodetected based on SRAM size.

maksimmasalski replied:
Yes, I agree with you. Seems that is why I received Shippable Error, because some devices can't support too many threads for work queues.

maksimmasalski replied:
@andyross Could you give an advice how it is better to implement? In configs documentation I can't find any SRAM sizes per platform.

maksimmasalski replied:
For now I just increased to CONFIG_MAX_THREAD_BYTES=3 to let Shippable passed and create 10 work queues for all platforms. I don't think if someone need more work queues in real application. Yes, that implementation is silly, because some platforms can handle more than 10 work queues... Maybe someone will want 50 work queues for his platform and, that is not tested and he will encounter with fatal error.

andyross reviewed (approved):
Some nitpicks on test cases that I'm pretty sure don't actually increase coverage.  Also a note on memory sizing.  But nothing that should block merge.

maksimmasalski commented on tests/kernel/workq/work_queue_api/src/main.c:
```diff
@@ -37,6 +50,154 @@ static struct k_sem sync_sema;
+ * - Then compare results.
+ * @ingroup kernel_workqueue_tests
+ */
+void test_sched_delayed_work_item(void)
```
I didn't find the test which checks that entered by user time for delay is correctly delayed in real program.

maksimmasalski replied:
I checked code and found test `test_delayed_work_submit_to_queue_thread` that is using function `tdelayed_work_submit_1` which tests the delayed time. But that function uses time got from a `k_delayed_work_remaining_get(w); ` as a default correct constant, and didn't try to verify that delayed time return from that function is correct and and least is in the limits of the real measurement. Also that test has some magic code line 
`timeout_ticks -= z_ms_to_ticks(15);` What means 15?
And I don't understand how that test verifies if the delayed time given by `k_delayed_work_remaining_get()` is really that time which was delayed by application in real life
 My test verifies that `k_delayed_work_remaining_get()` returned the correct number and that number is in the limits of the measured time using `k_cycle_get_32(); ` Unfortunately my test can't verify 100% that user defined delay of 100ms was in real life 100ms, because different arch use different timers, but it can definitely verify that delay in real life was 100ms +- delta 10ms.

I can keep old test, but I think necessary to modify it too. At least remove that magic line of code `timeout_ticks -= z_ms_to_ticks(15);` or make constant like I did with my delta constant.

andyross replied:
Heh, this is fine.  I didn't mean to send you crawling over the test suite (and, yes, a lot of our test code is of questionable quality).  There's no problem with adding a small and simple case for an obvious condition that isn't already tested in an obvious way.

maksimmasalski commented on tests/kernel/workq/work_queue_api/src/main.c:
```diff
@@ -37,6 +50,154 @@ static struct k_sem sync_sema;
+ * - If semaphore was obtained successfully, test passed.
+ * @ingroup kernel_workqueue_tests
+ */
+void test_work_item_supplied_with_func(void)
```
I checked tests again, and I think that test idea is already been tested by `test_work_submit_handler()`
I just wanted to create separate test for that feature.

Commit:
tests: resolve shippable errors

During previous commit Shippable let me know about errors
for some boards, resolved them.

Signed-off-by: Maksim Masalski <maksim.masalski@intel.com>

Commit:
tests: rebase with 2.3 release changes

Rebase using the latest master changes for 2.3 release

Signed-off-by: Maksim Masalski <maksim.masalski@intel.com>

maksimmasalski commented:
@nashif @andrewboie I'm looking for code review for that PR. Andy already gave comments.

sjg20 commented on tests/kernel/workq/work_queue_api/src/main.c:
```diff
@@ -37,6 +50,154 @@ static struct k_sem sync_sema;
+void test_sched_delayed_work_item(void)
+{
+	s32_t ms_remain, ms_spent, start_time, stop_time, cycles_spent;
+
```
drop blank line?

sjg20 commented on tests/kernel/workq/work_queue_api/src/main.c:
```diff
@@ -37,6 +50,154 @@ static struct k_sem sync_sema;
+	k_sem_reset(&sync_sema);
+
+	/* TESTPOINT: init delayed work to be processed */
+	/* only after specific period of time */
```
multi-line comment style

sjg20 commented on tests/kernel/workq/work_queue_api/src/main.c:
```diff
@@ -306,16 +308,25 @@ static void tdelayed_work_submit_1(struct k_work_q *work_q,
+	printk("\ntick_to_ms %lld\n\n", k_ticks_to_ms_floor64(timeout_ticks));
 
-	zassert_true(time_remaining >= k_ticks_to_ms_floor64(timeout_ticks),
+	zassert_true(time_remaining >= tick_to_ms,
```
join lines?
Same on 320

Commit:
tests: make code style changes according with review comments

Made light code style changes.

Signed-off-by: Maksim Masalski <maksim.masalski@intel.com>
